ctmap/gc: continue interval time on partial GC pass.

tommyp1ckles opened the pull request:
When {ipv4,ipv6} specific signals are sent from the datapath to indicate conditions such as high ctmap or snat pressure we may invoke a ipv4 or ipv6 *only* gc pass.
Currently this will also reset the gc interval by recomputing the next interval.

Under a situation where a lot of ipv4/ipv6 specific signals are being sent this may a starvation condition such that the other family never has a chance to be run since before the interval time can finish the loop is woken up by another signal.

To prevent this, if the GC was a "partial" pass (i.e. not including all *enabled* ip families), then we will continue waiting from where the previous timer left off.  Following a full GC pass we will just recompute the interval as normal (doesn't matter if the full pass was caused by the timer or a signal).

Addresses: #40820

aditighag commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -110,8 +110,23 @@ func New(params parameters) *GC {
+	// enabled IP families.
+	// Partial passes can happen as a result of datapath
+	// signals on particular families (i.e. ct or nat).
+	return ipv4 == gc.ipv4 && ipv6 == gc.ipv6
```
Umm... This is valid only for dual clusters, no?

tommyp1ckles replied:
Say `gc.ipv4` is true, and `gc.ipv6` (indicating the cluster is not dualstack), then the passed ipv4/ipv6 will be true,false so this will return true (which is expected... as this is a "Full GC" pass in the case of a single-stack cluster).

tommyp1ckles replied:
Just as a note, this clause was copied from below.  I wanted to put it into a separate function to improve readability.

aditighag commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -178,14 +195,38 @@ func (gc *GC) Enable() {
+				} else {
+					// partial pass, but the new interval is less than any leftover ttl so
+					// we cache this as well.
+					cachedGCInterval = interval
```
This can be simplified - 

```
forceInterval := max(0, time.Until(forceFullGCTTL))
if forceInterval < interval {
    interval = forceInterval
} else {
    cachedGCInterval = interval
}
```

tommyp1ckles replied:
Oh! missed the memo on there finally being built in min/max funcs in Go.

aditighag commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -251,13 +292,15 @@ func (gc *GC) Enable() {
-		// Not supporting BPF map pressure for local CT maps as of yet.
-		ctmap.CalculateCTMapPressure(gc.controllerManager, gc.metricsRegistry, ctmap.GlobalMaps(gc.ipv4, gc.ipv6)...)
-	}()
+	if runMapPressureDaemon {
```
How's this change related to the issues this PR is trying to fix?

tommyp1ckles replied:
Motivation: With this change, I wanted to add some real test coverage on the highLevel enable function, without having to worry about creating/mocking all the underlying map infrastructure.   This was specifically for the reasons we discussed - i.e. the control for for the GC loop getting too complicated so I wanted to be able to write some tests that _just_ check that logic.

So, before this would run a controller to do ctmap pressure iterate/counting, to avoid having to make all that mockable or create the maps I decided to just have it be optional when calling from the internal `gc.enable(...)` function and enabled by default for the exported `gc.Enable(...)` (i.e. similar to how I allow for passing a custom `doGC` to `enable`.

tommyp1ckles replied:
I'll put the testing refactors in a seperate commit however.

tommyp1ckles replied:
@aditighag I broke this commit into 3:

1) Do the Enable/enable refactor to prep for our unit test changes.
2) Implement the fix
3) Add unit test

aditighag commented on pkg/maps/ctmap/gc/gc_test.go:
```diff
@@ -0,0 +1,109 @@
+
+// TestGCEnable tests the overall *flow* of ctmap GC scheduling
+// while abstracting out the underlying garbage collection logic.
+func TestGCEnable(t *testing.T) {
```
It would be great to cover a couple of more tests cases as we've had many regressions in this code path:

- Receive both IPv4 and IPv6 signals
- The current test case is only testing low delete ratio (higher interval). It'll be good to also test a higher delete ratio (short interval).
- Receive IPv6 signals only (optional)

tommyp1ckles replied:
yup, good call, ill add extra coverage.

tommyp1ckles replied:
> The current test case is only testing low delete ratio (higher interval). It'll be good to also test a higher delete ratio (short interval).

note: I wanted to avoid adding this case, as testing relative "long vs short" durations would be more difficult to do than just checking if event "X" has happened within some reasonable timeout.

tommyp1ckles replied:
I think those test cases would be better covered by the GetInerval tests in package ctmap (although, imo this should probably be moved to package gc, alongside with its tests).

Basically, we probably don't want to actually have to wait in the test to see if it waited longer, but rather just check that GetDuration computes those times as we would expect.

aditighag replied:
>I think those test cases would be better covered by the GetInerval tests in package ctmap

I would like us to get more confidence on the changes in `runGC`. Higher delete ration => shorter interval
Not a blocking comment, but since we are early in the dev cycle, we don't have too much time pressure to ship the fix so it'll be good to work on the test coverage.

tommyp1ckles replied:
@aditighag Added a test for the interval "ratchet" behavior: https://github.com/cilium/cilium/pull/41258/commits
This test is really showing the difficulty of this code, definitely think we need to look at refactoring this but this should provide some assurances in not having regressions here before/after the changes.

Note: This runs successfully before and after the gc.Enable changes.

aditighag reviewed (changes requested):
The `runGC` function is getting overloaded in terms of functionalities, and as a result, carrying a lot of bookkeeping state. But I understand that the general refactoring is out of scope of this PR as we would want to backport this fix. 
Left comments around code simplification and a couple of more test cases.

tommyp1ckles committed:
ctmap: refactor cached GC interval out of ctmap package.

In subsequent commits we will want to make decisions on when
to set the cached GC interval depending on whether both enabled
ip families have had their GC run or just one (of two).

To do this, we need to move the cachedGCInterval out of the ctmap
package and make this explicit in the GetInterval function.

To prep for that we fix the tests and set the cached interval
outside of GetInterval.

Addresses: #40820

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ctmap/gc: refactor gc.Enable to allow for testing gc scheduling flow

Current implementation is tied to the real runGC function that requires
real bpf maps etc to run. As the enable functionality becomes more
complex we need to be able to write tests targeting the core schedule
flow of running GC - separate from actual GC implementation logic.

To help us write more focused tests in subsequent commits, we maintain
the same exported Enable function API but underneath provide an internal
enable() (called by Enable with default params) that allows us to pass
a custom runGC func as well as disable the map pressure counter
controller.

Specifically, subsequent commits will add unit tests covering different
scenarios for ipv4/ipv6 partia/full signal handling.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ctmap/gc: continue interval time on partial GC pass.

When {ipv4,ipv6} specific signals are sent from the datapath
to indicate conditions such as high ctmap or snat pressure
we may invoke a ipv4 or ipv6 *only* gc pass.
Currently this will also reset the gc interval by recomputing
the next interval.

Under a situation where a lot of ipv4/ipv6 specific signals
are being sent this may a starvation condition such that the
other family never has a chance to be run since before the
interval time can finish the loop is woken up by another signal.

To improve this, if the GC was a "partial" pass (i.e. not including
all *enabled* ip families), then we will continue waiting from where
the previous timer left off.  Following a full GC pass we will just
recompute the interval as normal (doesn't matter if the full pass
was caused by the timer or a signal).

Fixes: #40820

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ctmap/gc: add unit test to check flow of partial/full gc runs.

* Partial: not all enabled ip families were gc'd.
* Full: all enabled ip families were gc'd.

With these definitions and refactors in previous commits to
allow decoupling runGC logic from the enable gc scheduling,
we add unit tests that validate the various signal -> gc pass
flows to verify that we no longer encounter starvation conditions.

Addresses: #40820

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

pchaigno commented on pkg/maps/ctmap/gc/gc.go:
```diff
@@ -191,12 +200,31 @@ func (gc *GC) enable(
 
 			// Mark the CT GC as over in each EP DNSZombies instance, if we did a *full* GC run
```
Nitpick: This comment probably needs to be moved 3 lines down.

pchaigno reviewed (commented):
Thanks for tackling this and for the neat commit history!

In a scenario where there's a lot more traffic for one family than the other, won't this cause us to run GC more often? Was the CPU impact measured?

tommyp1ckles committed:
ctmap: allow adjusting min gc interval and rounding granularity.

Cilium uses the default values here, but for testing it would be
very useful to allow decreasing these to avoid having overly long
running tests.

Subsequent commits will use these changes to modify these in
testing to speed up test times.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
ctmap/gc: add test that validates gc interval "ratchet" behavior.

CTMap GC uses both previous interval times and delete ratios
to compute its next interval.

This commit adds a test that checks that somewhat complex behavior
and prevents future regressions.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

aditighag reviewed (approved):
Thanks for the revisions, LGTM!
